lighthouse-core/audits/seo/structured-data-automatic.js

```diff
@@ -0,0 +1,173 @@
+    }
+
+    const validatedSnippets = await Promise.all(jsonLDElements.map(async (element) => {
+      // We don't want to show empty lines around the snippet
+      const content = (element.content || '').trim();
+
+      return {
+        devtoolsNodePath: element.devtoolsNodePath,
+        content,
+        errors: await validateJsonLD(content),
+      };
+    }));
+    // Show invalid snippets at the top
+    validatedSnippets.sort((a, b) => {
+      return b.errors.length - a.errors.length;
+    });
+
+    const renderedSnippets = await Promise.all(
+      validatedSnippets.map(async (snippetWithErrors) => {
+        return renderValidatedSnippet(snippetWithErrors);
+      })
+    );
+    const details = Audit.makeListDetails(renderedSnippets);
+
+    const invalidSnippets = validatedSnippets.filter(vs => vs.errors.length > 0);
+    const validSnippets = validatedSnippets.filter(vs => vs.errors.length === 0);
+    const displayValue = str_(UIStrings.displayValue, {
+      validSnippetProportion: validSnippets.length / jsonLDElements.length,
+    });
+
+    return {
+      rawValue: invalidSnippets.length === 0,
+      score: invalidSnippets.length === 0 ? 1 : 0,
+      details,
+      displayValue,
+    };
+  }
+}
+
+/**
+ * @param {{content: string, devtoolsNodePath: string, errors: LH.StructuredData.ValidationError[]}} validatedSnippet
+ */
+function renderValidatedSnippet(validatedSnippet) {
+  const {content, devtoolsNodePath, errors} = validatedSnippet;
+
+  let parsedContent;
+  let topLevelType;
+  let topLevelName;
+  try {
+    parsedContent = JSON.parse(content);
+    topLevelType = parsedContent['@type'];
+    topLevelName = parsedContent['name'];
+  } catch (err) {}
+
+  let title = '';
+  if (topLevelName && topLevelType) {
+    title = `${topLevelType}: ${topLevelName}`;
+  } else if (topLevelType) {
+    title = `@type ${topLevelType}`;
+  } else {
+    title = 'Invalid JSON-LD element';
+  }
+  title += ` (${errors.length} Error${errors.length !== 1 ? 's' : ''})`;
+
+  /** @type LH.Audit.Details.NodeValue */
+  const node = {
+    type: 'node',
+    path: devtoolsNodePath,
+    snippet: `<script type="application/ld+json">`,
+  };
+
+  const {lineMessages, generalMessages} = getErrorMessages(errors);
+
+  return Audit.makeSnippetDetails({
+    content: parsedContent ? JSON.stringify(parsedContent, null, 2) : content,
+    title,
+    lineMessages,
+    generalMessages,
+    node,
+  });
+}
+
+/**
+ * @param {Array<LH.StructuredData.ValidationError>} errors
+ */
+function getErrorMessages(errors) {
+  /** @type {LH.Audit.Details.SnippetValue['lineMessages']} */
+  const lineMessages = [];
+  /** @type {LH.Audit.Details.SnippetValue['generalMessages']} */
+  const generalMessages = [];
+  errors.forEach(({
+    message, lineNumber, validTypes, validator,
+  }) => {
+    if (validTypes && validator === 'schema-org') {
+      const typeStrings = validTypes.map(type => {
+        const typeNameMatch = type.match(/[\w]+$/);
```

Comment:
this feels unfortunate we have to do this regex business here

we control the library is there another way to feed this more directly, or you think it still makes sense to do this parsing?

Reply:
Let's do the regex business in sd-validator so that `validTypes` is an object with `uri` and `name` properties? I don't think we can avoid the regex altogether.

Reply:
oh yeah sg! that's what I meant, just that having to parse out with a regex *post* sd library feels rough and I'm assuming if others ever wanted to use this they could want something similar